lib/Jit/jitoptions.cpp

```diff
@@ -50,11 +58,52 @@ void freeStringConfigValue(ICorJitInfo *CorInfo, UTF16CharT *Value) {
 }
 
 JitOptions::JitOptions(LLILCJitContext &Context) {
-  // Set 'IsAltJit' based on environment information.
-  IsAltJit = queryIsAltJit(Context);
+
+  if (!StaticOptionsInitialized) {
+    StaticDumpLevel = ::DumpLevel::NODUMP;
+
+    char16_t *LevelWStr =
+      getStringConfigValue(Context.JitInfo, UTF16("DUMPLLVMIR"));
+    if (LevelWStr) {
+      std::unique_ptr<std::string> Level = Convert::utf16ToUtf8(LevelWStr);
+      std::transform(Level->begin(), Level->end(), Level->begin(), ::toupper);
+      if (Level->compare("VERBOSE") == 0) {
+        StaticDumpLevel = ::DumpLevel::VERBOSE;
+
+      } else if (Level->compare("SUMMARY") == 0) {
+        StaticDumpLevel = ::DumpLevel::SUMMARY;
+      }
+    }
+    if (Context.Flags & CORJIT_FLG_PREJIT) {
+      InitMethodSet(Context, AltJitNgenMethodSet, UTF16("AltJitNgen"));
+      // Set up AltJitNgen set to be used for AltJit test.
+      AltJitPtr = &AltJitNgenMethodSet;
+    } else {
+      InitMethodSet(Context, AltJitMethodSet, UTF16("AltJit"));
+      // Set up AltJit set to be use for AltJit test.
+      AltJitPtr = &AltJitMethodSet;
+    }
+    InitMethodSet(Context, ExcludeMethodSet, UTF16("AltJitExclude"));
+    InitMethodSet(Context, BreakMethodSet, UTF16("AltJitBreak"));
+    InitMethodSet(Context, MSILMethodSet, UTF16("AltJitMSILDump"));
+    InitMethodSet(Context, LLVMMethodSet, UTF16("AltJitLLVMDump"));
+    InitMethodSet(Context, EntrySizeMethodSet, UTF16("AltJitCodeRangeDump"));
+    StaticOptionsInitialized = true;
+  }
+
 
   // Set dump level for this JIT invocation.
-  DumpLevel = queryDumpLevel(Context);
+  DumpLevel = StaticDumpLevel;
+  const char *ClassName = nullptr;
+  const char *MethodName = 
+    Context.JitInfo->getMethodName(Context.MethodInfo->ftn, &ClassName);
+  PCCOR_SIGNATURE PCSig = Context.MethodInfo->args.pSig;
+  IsAltJit = AltJitPtr->contains(MethodName, ClassName, PCSig);
+  IsExcludeMethod = ExcludeMethodSet.contains(MethodName, ClassName, PCSig);
+  IsBreakMethod = BreakMethodSet.contains(MethodName, ClassName, PCSig);
+  IsMSILDumpMethod = MSILMethodSet.contains(MethodName, ClassName, PCSig);
+  IsLLVMDumpMethod = LLVMMethodSet.contains(MethodName, ClassName, PCSig);
+  IsEntrySizeMethod = EntrySizeMethodSet.contains(MethodName, ClassName, PCSig);
```

Comment:
I don't particular like this approach nor do I buy the efficiency argument.  The separate routines were to allow for separate query for each that was easily understandable on it's own.  If you look at the RyuJIT handling you can see it's organically grown into the very large convolution of code.  Keeping each option query separate allows for much easier separation of the logic.  Having an isInitialized per methodset allows for the method set to be used more flexabily while preserving the (simple) thread safety.  My main complaint is that the options are hand inlined into the constructor and when the number of options them grows into the double or tipple digits this is going to make for a really confusing method.  We should probably catch up in person.